akka-persistence-testkit/src/main/scala/akka/persistence/testkit/internal/Unpersistent.scala

```diff
@@ -0,0 +1,483 @@
+
+    sendSignal(RecoveryCompleted)
+
+    override def onMessage(cmd: Command): Behavior[Command] = {
+      var shouldUnstash = false
+      var shouldStop = false
+
+      def persistState(st: State): Unit = {
+        sequenceNr += 1
+        onPersist(st, sequenceNr, tag)
+        state = st
+      }
+
+      @tailrec
+      def applyEffects(curEffect: EffectImpl[State], sideEffects: immutable.Seq[SideEffect[State]]): Unit =
+        curEffect match {
+          case CompositeEffect(eff: EffectImpl[_], se) =>
+            applyEffects(eff.asInstanceOf[EffectImpl[State]], se ++ sideEffects)
+
+          case Persist(st) =>
+            persistState(st)
+            sideEffect(sideEffects)
+
+          case _: PersistNothing.type | _: Unhandled.type =>
+            sideEffect(sideEffects)
+
+          case _: Stash.type =>
+            stashedCommands.append(cmd)
+            sideEffect(sideEffects)
+
+          case _ =>
+            context.log.error("Unexpected effect, stopping")
+            Behaviors.stopped
+        }
+
+      def sideEffect(sideEffects: immutable.Seq[SideEffect[State]]): Unit =
+        sideEffects.iterator.foreach { effect =>
+          effect match {
+            case _: Stop.type       => shouldStop = true
+            case _: UnstashAll.type => shouldUnstash = true
+            case cb: Callback[_]    => cb.sideEffect(state)
+          }
+        }
+
+      applyEffects(commandHandler(state, cmd).asInstanceOf[EffectImpl[State]], Nil)
+
+      if (shouldStop) Behaviors.stopped
+      else if (shouldUnstash && stashedCommands.nonEmpty) {
+        val numStashed = stashedCommands.length
+        val thisWrappedBehavior = this
+
+        Behaviors.setup { _ =>
+          Behaviors.withStash(numStashed) { stash =>
+            stashedCommands.foreach { sc =>
+              stash.stash(sc)
+              () // explicit discard
+            }
+            stashedCommands.remove(0, numStashed)
+            stash.unstashAll(thisWrappedBehavior)
+          }
+        }
+      } else this
+    }
+  }
+}
+
+class PersistenceProbeImpl[T] {
+  type Element = (T, Long, Set[String])
+
+  val queue = new ConcurrentLinkedQueue[Element]()
+
+  def persist(elem: Element): Unit = { queue.offer(elem); () }
+
+  def rawExtract(): Element =
+    queue.poll() match {
+      case null => throw new AssertionError("No persistence effects in probe")
+      case elem => elem
+    }
+
+  def asScala: scaladsl.PersistenceProbe[T] =
+    new scaladsl.PersistenceProbe[T] {
+      import scaladsl.PersistenceProbe
+
+      def drain(): Seq[Element] = {
+        @annotation.tailrec
+        def iter(acc: List[Element]): List[Element] = {
+          val elem = queue.poll()
+          if (elem == null) acc else iter(elem :: acc)
+        }
+
+        iter(Nil).reverse
+      }
+
+      def extract(): Element = rawExtract()
+
+      def expectPersistedType[S <: T : ClassTag](): (S, Long, Set[String]) = {
```

Comment:
I think a case class with names for the types would make more sense than a tuple here as well.